File: Tps/Tp4/TP04Q07 - Hash Indireta/Main.java
import java.util.*;
import java.text.ParseException;
import java.text.SimpleDateFormat;

class Main {
    static double inicio = System.currentTimeMillis();
    static int comp = 0;
    static Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) {
        Questao7();
        scanner.close();
    }

    // ---------------------------------------------Questoes---------------------------------------------
    static void Questao7() {
        String x;
        boolean fim = false;
        ArrayList<Personagem> personagens = new ArrayList<>();
        personagens = readArq();
        HashIndiretoLista tabelaHash = new HashIndiretoLista(21);
        while (fim == false) {
            x = scanner.nextLine();
            if (x.equals("FIM")) {
                fim = true;
            } else {
                for (Personagem personagem : personagens) {
                    if (personagem.getId().equals(x)) {
                        tabelaHash.inserirInicio(personagem);
                    }
                }
            }
        }
        fim = false;
        while (fim == false) {
            x = scanner.nextLine();
            if (x.equals("FIM")) {
                fim = true;
            } else {
                System.out.print(x);
                System.out.println(tabelaHash.pesquisar(x) ? " SIM" : " NAO");
            }
        }
        criaArqLog();
    }

    static public class HashIndiretoLista {
        Lista tabela[];
        int tamanho;        

        public HashIndiretoLista() {
            this(21);
        }

        public HashIndiretoLista(int tamanho) {
            this.tamanho = tamanho;
            tabela = new Lista[tamanho];
            for (int i = 0; i < tamanho; i++) {
                tabela[i] = new Lista();
            }
        }

        public int h(String elemento) {
            int soma=0;
            for(int i=0;i<elemento.length();i++){
                soma+=elemento.charAt(i);
            }
            return soma % tamanho;
        }

        boolean pesquisar(String elemento) {
            int pos = h(elemento);            
            return tabela[pos].pesquisar(elemento,pos);
        }

        public void inserirInicio(Personagem elemento) {
            int pos = h(elemento.getName());
            tabela[pos].inserirInicio(elemento);
        }
        
    }

    static class Celula {
        public Personagem elemento; // Elemento inserido na celula.
        public Celula prox; // Aponta a celula prox.

        Celula(Personagem elemento) {
            this.elemento = elemento;
            this.prox = null;
        }

        Celula(Personagem elemento, Celula prox) {
            this.elemento = elemento;
            this.prox = prox;
        }
    }

    static public class Lista {
        private Celula primeiro; // Primeira celula: SEM elemento valido.
        private Celula ultimo; // Ultima celula: COM elemento valido.

        public Lista() {
            primeiro = new Celula(null);
            ultimo = primeiro;
        }

        public boolean pesquisar(String x,int pos) {
            boolean retorno = false;
            for (Celula i = primeiro.prox; i != null; i = i.prox) {
                comp++;
                if (i.elemento.getName().equals(x)) {
                    retorno = true;
                    i = ultimo;
                }
            }
            if (retorno == true){
                System.out.print(" (pos: "+pos+")");
            }
            return retorno;
        }

        public void inserirInicio(Personagem elemento) {
            Celula tmp = new Celula(elemento);
            tmp.prox = primeiro.prox;
            primeiro.prox = tmp;
            comp++;
            if (primeiro == ultimo) {
                ultimo = tmp;
            }
            tmp = null;
        }

    }

    // ---------------------------------------------CriaArqLog---------------------------------------------
    static void criaArqLog() {
        Arq.openWrite("matricula_hashIndireta.txt");
        Arq.print("694520\t" + (System.currentTimeMillis() - (inicio / 1000)) + "\t" + comp);
        Arq.close();
    }

    // ---------------------------------------------readArq---------------------------------------------
    public static ArrayList<Personagem> readArq() {
        String x;
        ArrayList<Personagem> personagens = new ArrayList<>();
        Arq.openRead("characters.csv");
        x = Arq.readLine();
        while (Arq.hasNext()) {
            x = Arq.readLine();
            Personagem personagem = new Personagem();
            x = x.replaceAll(";;", "; ;").replaceAll("\\[", "{").replaceAll("]", "}");
            personagem.ler(x);
            personagens.add(personagem);
        }
        Arq.close();
        return personagens;
    }

    // ---------------------------------------------Classe_Personagem---------------------------------------------
    static class Personagem {
        private String id;
        private String name;
        private List<String> alternate_names = new ArrayList<String>();
        private String house;
        private String ancestry;
        private String species;
        private String patronus;
        private boolean hogwartsStaff;
        private String hogwartsStudent;
        private String actorName;
        private boolean alive;
        private Date dateOfBirth;
        private int yearOfBirth;
        private String eyeColor;
        private String gender;
        private String hairColour;
        private boolean wizard;

        public Personagem() {
            id = "";
            name = "";
            alternate_names = new ArrayList<String>();
            house = "";
            ancestry = "";
            species = "";
            patronus = "";
            hogwartsStaff = false;
            hogwartsStudent = "";
            actorName = "";
            alive = false;
            dateOfBirth = new Date();
            yearOfBirth = 0;
            eyeColor = "";
            gender = "";
            hairColour = "";
            wizard = false;
        }

        public Personagem(String id, String name, ArrayList<String> alternate_names, String house, String ancestry,
                String species,
                String patronus, boolean hogwartsStaff,
                String hogwartsStudent, String actorName, boolean alive, Date dateOfBirth, int yearOfBirth,
                String eyeColor, String gender, String hairColour, boolean wizard) {
            this.id = id;
            this.name = name;
            this.alternate_names = new ArrayList<String>();
            for (String temp : alternate_names) {
                this.alternate_names.add(temp);
            }
            this.house = house;
            this.ancestry = ancestry;
            this.species = species;
            this.patronus = patronus;
            this.hogwartsStaff = hogwartsStaff;
            this.hogwartsStudent = hogwartsStudent;
            this.actorName = actorName;
            this.alive = alive;
            this.dateOfBirth = new Date();
            this.yearOfBirth = yearOfBirth;
            this.eyeColor = eyeColor;
            this.gender = gender;
            this.hairColour = hairColour;
            this.wizard = wizard;
        }

        // gets
        public String getId() {
            return id;
        }

        public String getName() {
            return name;
        }

        public String getAlternateNames() {
            StringBuilder sb = new StringBuilder();
            for (String name : alternate_names) {
                sb.append(name).append(", ");
            }
            // Remover a última vírgula e espaço, se houver
            if (sb.length() > 0) {
                sb.setLength(sb.length() - 2);
            }
            return sb.toString();
        }

        public String getHouse() {
            return house;
        }

        public String getAncestry() {
            return ancestry;
        }

        public String getSpecies() {
            return species;
        }

        public String getPatronus() {
            return patronus;
        }

        public boolean getHogwartsStaff() {
            return hogwartsStaff;
        }

        public String getHogwartsStudent() {
            return hogwartsStudent;
        }

        public String getActorName() {
            return actorName;
        }

        public boolean getAlive() {
            return alive;
        }

        public Date getDateOfBirth() {
            return dateOfBirth;
        }

        public int getYearOfBirth() {
            return yearOfBirth;
        }

        public String getEyeColor() {
            return eyeColor;
        }

        public String getGender() {
            return gender;
        }

        public String getHairColour() {
            return hairColour;
        }

        public boolean getWizard() {
            return wizard;
        }

        // sets
        public void setId(String id) {
            this.id = id;
        }

        public void setName(String name) {
            if (name.equals(" ")) {
                name = "";
            }
            this.name = name;
        }

        public void setAlternateNames(String nomesAlterados) {
            if (nomesAlterados.charAt(0) == ' ') {
                StringBuilder nomesAlteradosNovo = new StringBuilder();
                for (int i = 1; i < nomesAlterados.length(); i++) {
                    nomesAlteradosNovo.append(nomesAlterados.charAt(i));
                }
                String x = nomesAlteradosNovo.toString();
                this.alternate_names.add(x);
            } else {
                this.alternate_names.add(nomesAlterados);
            }

        }

        public void setHouse(String house) {
            if (house.equals(" ")) {
                house = "";
            }
            this.house = house;
        }

        public void setAncestry(String ancestry) {
            if (ancestry.equals(" ")) {
                ancestry = "";
            }
            this.ancestry = ancestry;
        }

        public void setSpecies(String species) {
            if (species.equals(" ")) {
                species = "";
            }
            this.species = species;
        }

        public void setPatronus(String patronus) {
            if (patronus.equals(" ")) {
                patronus = "";
            }
            this.patronus = patronus;
        }

        public void setHogwartsStaff(boolean hogwartsStaff) {
            this.hogwartsStaff = hogwartsStaff;
        }

        public void setHogwartsStudent(String hogwartsStudent) {
            if (hogwartsStudent.equals(" ")) {
                hogwartsStudent = "";
            }
            this.hogwartsStudent = hogwartsStudent;
        }

        public void setActorName(String actorName) {
            if (actorName.equals(" ")) {
                actorName = "";
            }
            this.actorName = actorName;
        }

        public void setAlive(boolean alive) {
            this.alive = alive;
        }

        public void setDateOfBirth(String dateOfBirth) {
            if (dateOfBirth.equals(" ")) {
                dateOfBirth = "";
            }
            String[] partes = dateOfBirth.split("-");
            StringBuilder temp = new StringBuilder();

            if (partes[0].length() < 2) {
                temp.append("0");
                temp.append(partes[0] + "-");
            } else {
                temp.append(partes[0] + "-");
            }
            if (partes[1].length() < 2) {
                temp.append("0");
                temp.append(partes[1] + "-");
            } else {
                temp.append(partes[1] + "-");
            }
            temp.append(partes[2]);
            String tempString = temp.toString();
            SimpleDateFormat formato = new SimpleDateFormat("dd-MM-yyyy");
            try {
                this.dateOfBirth = formato.parse(tempString);
            } catch (ParseException e) {

            }
        }

        public void setYearOfBirth(String yearOfBirth) {
            int anoNascimento = 0;
            try {
                anoNascimento = Integer.parseInt(yearOfBirth);
                this.yearOfBirth = anoNascimento;
            } catch (Exception e) {
                this.yearOfBirth = 0;
            }

        }

        public void setEyeColor(String eyeColor) {
            if (eyeColor.equals(" ")) {
                eyeColor = "";
            }
            this.eyeColor = eyeColor;
        }

        public void setGender(String gender) {
            if (gender.equals(" ")) {
                gender = "";
            }
            this.gender = gender;
        }

        public void setHairColour(String hairColour) {
            if (hairColour.equals(" ")) {
                hairColour = "";
            }
            this.hairColour = hairColour;
        }

        public void setWizard(boolean wizard) {
            this.wizard = wizard;
        }

        // clone
        public Personagem clone() {
            Personagem clone = new Personagem();
            clone.id = this.id;
            clone.name = this.name;
            clone.alternate_names = this.alternate_names;
            clone.house = this.house;
            clone.ancestry = this.ancestry;
            clone.species = this.species;
            clone.patronus = this.patronus;
            clone.hogwartsStaff = this.hogwartsStaff;
            clone.hogwartsStudent = this.hogwartsStudent;
            clone.actorName = this.actorName;
            clone.alive = this.alive;
            clone.dateOfBirth = this.dateOfBirth;
            clone.yearOfBirth = this.yearOfBirth;
            clone.eyeColor = this.eyeColor;
            clone.gender = this.gender;
            clone.hairColour = this.hairColour;
            clone.wizard = this.wizard;
            return clone;
        }

        public void ler(String x) {
            // 0 = id 1 = name 2 = alternate_names 3 = house
            // 4 = ancestry 5 = species 6 = patronus 7 = hogwartsStaff
            // 8 = hogwartsStudent 9 = actorName 10 = alive 12 = dateOfBirth
            // 13 = yearOfBirth 14 = eyeColor 15 = gender 16 = hairColour
            // 17 = wizard

            String[] partes = x.split(";");

            setId(partes[0]);
            setName(partes[1]);
            if (partes[2].length() >= 1) {
                partes[2] = partes[2].replaceAll("'", "");
                String[] nomesAlternados = partes[2].split(",");
                for (int i = 0; i < nomesAlternados.length; i++) {
                    setAlternateNames(nomesAlternados[i]);
                }
            } else {
                setAlternateNames("");
            }
            setHouse(partes[3]);
            setAncestry(partes[4]);
            setSpecies(partes[5]);
            setPatronus(partes[6]);

            if (partes[7].equals("FALSO")) {
                setHogwartsStaff(false);
            } else if (partes[7].equals("VERDADEIRO")) {
                setHogwartsStaff(true);
            }

            if (partes[8].equals("FALSO")) {
                setHogwartsStudent("false");
            } else if (partes[8].equals("VERDADEIRO")) {
                setHogwartsStudent("true");
            }

            setActorName(partes[9]);

            if (partes[10].equals("FALSO")) {
                setAlive(false);
            } else if (partes[10].equals("VERDADEIRO")) {
                setAlive(true);
            }

            setDateOfBirth(partes[12]);

            setYearOfBirth(partes[13]);

            setEyeColor(partes[14]);
            setGender(partes[15]);
            setHairColour(partes[16]);

            if (partes[17].equals("FALSO")) {
                setWizard(false);
            } else if (partes[17].equals("VERDADEIRO")) {
                setWizard(true);
            }
        }

        public void imprimir(int pos) {
            SimpleDateFormat formatoSaida = new SimpleDateFormat("dd-MM-yyyy");
            String dataFormatada = formatoSaida.format(this.dateOfBirth);
            System.out.print("[" + pos + " ## " + this.id + " ## " + this.name + " ## " + this.getAlternateNames()
                    + " ## " +
                    this.house + " ## " + this.ancestry + " ## " + this.species + " ## " + this.patronus + " ## "
                    + this.hogwartsStaff
                    + " ## " + this.hogwartsStudent + " ## " + this.actorName + " ## " + this.alive + " ## "
                    + dataFormatada + " ## " + this.yearOfBirth
                    + " ## " + this.eyeColor + " ## " + this.gender + " ## " + this.hairColour + " ## " + this.wizard
                    + "]\n");
        }

    }
}

File: Tps/Tp4/TP04Q07 - Hash Indireta/main.c
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <time.h>

int comp=0;
clock_t startTime;
clock_t endTime;

void startTimer() {
    startTime = clock();
}
void endTimer() {
    endTime = clock();
}


typedef struct
{
    char Lista[500];
} Lista;

typedef struct
{
    char id[200];
    char name[200];
    Lista alternate_names;
    char house[200];
    char ancestry[200];
    char species[200];
    char patronus[200];
    char hogwartsStaff[10];
    char hogwartsStudent[10];
    char actorName[200];
    char alive[10];
    char dateOfBirth[200];
    int yearOfBirth;
    char eyeColour[200];
    char gender[200];
    char hairColor[200];
    char wizard[10];

} Personagem;

Personagem constr(char id[], char name[], char alternate_names[], char house[], char ancestry[], char species[], char patronus[], char hogwartsStaff[], char hogwartsStudent[], char actorName[], char alive[], char dateOfBirth[],
                  int yearOfBirth, char eyeColour[], char gender[], char hairColor[], char wizard[])
{
    Personagem personagem;

    strcpy(personagem.id, id);
    strcpy(personagem.name, name);
    strcpy(personagem.alternate_names.Lista, alternate_names);
    strcpy(personagem.house, house);
    strcpy(personagem.ancestry, ancestry);
    strcpy(personagem.patronus, patronus);
    strcpy(personagem.species, species);
    strcpy(personagem.hogwartsStaff, hogwartsStaff);
    strcpy(personagem.hogwartsStudent, hogwartsStudent);
    strcpy(personagem.actorName, actorName);
    strcpy(personagem.alive, alive);
    strcpy(personagem.dateOfBirth, dateOfBirth);
    personagem.yearOfBirth = yearOfBirth;
    strcpy(personagem.eyeColour, eyeColour);
    strcpy(personagem.gender, gender);
    strcpy(personagem.hairColor, hairColor);
    strcpy(personagem.wizard, wizard);

    return personagem;
}

Personagem constrVazio()
{
    Personagem personagem;
    strcpy(personagem.id, "");
    strcpy(personagem.name, "");
    strcpy(personagem.alternate_names.Lista, "");
    strcpy(personagem.house, "");
    strcpy(personagem.ancestry, "");
    strcpy(personagem.patronus, "");
    strcpy(personagem.species, "");
    strcpy(personagem.hogwartsStaff, "");
    strcpy(personagem.hogwartsStudent, "");
    strcpy(personagem.actorName, "");
    strcpy(personagem.alive, "");
    strcpy(personagem.dateOfBirth, "");
    personagem.yearOfBirth = 0;
    strcpy(personagem.eyeColour, "");
    strcpy(personagem.gender, "");
    strcpy(personagem.hairColor, "");
    strcpy(personagem.wizard, "");

    return personagem;
}

// getters
char *getId(Personagem *personagem) { return personagem->id; }
char *getName(Personagem *personagem) { return personagem->name; }
char *getAlternate_names(Personagem *personagem) { return personagem->alternate_names.Lista; }
char *getHouse(Personagem *personagem) { return personagem->house; }
char *getAncestry(Personagem *personagem) { return personagem->ancestry; }
char *getSpecies(Personagem *personagem) { return personagem->species; }
char *getPatronus(Personagem *personagem) { return personagem->patronus; }
char *getHogwartsStaff(Personagem *personagem) { return personagem->hogwartsStaff; }
char *getHogwartsStudent(Personagem *personagem) { return personagem->hogwartsStudent; }
char *getActorName(Personagem *personagem) { return personagem->actorName; }
char *getAlive(Personagem *personagem) { return personagem->alive; }
char *getDateOfBirth(Personagem *personagem) { return personagem->dateOfBirth; }
int getYearOfBith(Personagem *personagem) { return personagem->yearOfBirth; }
char *getEyeColour(Personagem *personagem) { return personagem->eyeColour; }
char *getGender(Personagem *personagem) { return personagem->gender; }
char *getHairColor(Personagem *personagem) { return personagem->hairColor; }
char *getWizard(Personagem *personagem) { return personagem->wizard; }

// setters
void setId(char *id, Personagem *personagem) { strcpy(personagem->id, id); }
void setName(char *name, Personagem *personagem) { strcpy(personagem->name, name); }
void setAlternate_names(char *alternate_names, Personagem *personagem) { strcpy(personagem->alternate_names.Lista, alternate_names); }
void setHouse(char *house, Personagem *personagem) { strcpy(personagem->house, house); }
void setAncestry(char *species, Personagem *personagem) { strcpy(personagem->species, species); }
void setSpecies(char *species, Personagem *personagem) { strcpy(personagem->species, species); }
void setPatronus(char *patronus, Personagem *personagem) { strcpy(personagem->patronus, patronus); }
void setHogwartsStaff(char *hogwartsStaff, Personagem *personagem) { strcpy(personagem->hogwartsStaff, hogwartsStaff); }
void setHogwartsStudent(char *hogwartsStudent, Personagem *personagem) { strcpy(personagem->hogwartsStudent, hogwartsStudent); }
void setActorName(char *actorName, Personagem *personagem) { strcpy(personagem->actorName, actorName); }
void setAlive(char *alive, Personagem *personagem) { strcpy(personagem->alive, alive); }
void setDateOfBirth(char *dateOfBirth, Personagem *personagem) { strcpy(personagem->dateOfBirth, dateOfBirth); }
void setYearOfBith(int yearOfBirth, Personagem *personagem) { personagem->yearOfBirth = yearOfBirth; }
void setEyeColour(char *eyeColour, Personagem *personagem) { strcpy(personagem->eyeColour, eyeColour); }
void setGender(char *gender, Personagem *personagem) { strcpy(personagem->gender, gender); }
void setHairColor(char *hairColor, Personagem *personagem) { strcpy(personagem->hairColor, hairColor); }
void setWizard(char *wizard, Personagem *personagem) { strcpy(personagem->wizard, wizard); }

// ler
char **ler(char line[])
{
    int tam_line = strlen(line);
    int start = 0;
    int count = 0;
    char **elementos = malloc(18 * sizeof(char *));
    if (elementos == NULL) {
        // Tratamento de erro
        exit(1);
    }

    for (int i = 0; i < 18; i++)
    {
        elementos[i] = malloc(500);
        if (elementos[i] == NULL) {
            // Tratamento de erro
            exit(1);
        }
    }

    for (int i = 0; i < tam_line; i++)
    {
        if (line[i] == ';' && line[i - 1] == ';')
        {
            strcpy(elementos[count], "");
            start = i + 1;
            count++;
        }
        else if (line[i] == ';')
        {
            strncpy(elementos[count], line + start, i - start);
            elementos[count][i - start] = '\0';
            start = i + 1;
            count++;
        }
    }

    strncpy(elementos[17], line + start, (tam_line - start) - 1);

    return elementos;
}

// imprimir
void imprimir(Personagem *personagem)
{
    printf("[%s ## %s ## %s ## %s ## %s ## %s ## %s ## %s ## %s ## %s ## %s ## %s ## %d ## %s ## %s ## %s ## %s]\n",
           personagem->id, personagem->name, personagem->alternate_names.Lista, personagem->house, personagem->ancestry, personagem->species, personagem->patronus,
           personagem->hogwartsStaff, personagem->hogwartsStudent, personagem->actorName, personagem->alive, personagem->dateOfBirth,
           personagem->yearOfBirth, personagem->eyeColour, personagem->gender, personagem->hairColor, personagem->wizard);
}
void removerCaractere(char *parte, char c)
{
    int i = 0;
    int j = 0;
    while (parte[i] != '\0')
    {
        if (parte[i] != c)
            parte[j++] = parte[i];
        i++;
    }
    parte[j] = '\0';
}
void preencherVetor(Personagem personagens[])
{
    FILE *arq;
    char line[1200];
    if ((arq = fopen("/tmp/characters.csv", "r")) != NULL)
    {
        int i = 0;
        fgets(line, 500, arq);
        while (fgets(line, 1200, arq) != NULL)
        {
            char **partes = ler(line);
            removerCaractere(partes[2], '\'');
            for (int i = 0; partes[2][i] != '\0'; i++)
            {
                if (partes[2][i] == '[')
                {
                    partes[2][i] = '{';
                }
                else if (partes[2][i] == ']')
                {
                    partes[2][i] = '}';
                }
            }
            if (strcmp(partes[7], "VERDADEIRO") == 0)
            {
                strcpy(partes[7], "true");
            }
            else
            {
                strcpy(partes[7], "false");
            }
            if (strcmp(partes[8], "VERDADEIRO") == 0)
            {
                strcpy(partes[8], "true");
            }
            else
            {
                strcpy(partes[8], "false");
            }
            if (strcmp(partes[10], "VERDADEIRO") == 0)
            {
                strcpy(partes[10], "true");
            }
            else
            {
                strcpy(partes[10], "false");
            }
            if (strcmp(partes[17], "VERDADEIRO") == 0)
            {
                strcpy(partes[17], "true");
            }
            else
            {
                strcpy(partes[17], "false");
            }
            personagens[i] = constr(partes[0], partes[1], partes[2], partes[3], partes[4],
                                    partes[5], partes[6], partes[7], partes[8], partes[9], partes[10],
                                    partes[12], atoi(partes[13]), partes[14], partes[15], partes[16], partes[17]);

            i++;
            free(partes);
        }
        fclose(arq);
    }
    else
    {
        printf("Erro ao abrir o arquivo");
    }
}

typedef struct Celula {
    Personagem elemento;
    struct Celula *prox;
} Celula;

typedef struct {
    Celula *cabeca;
    int count;
} tabelaHash;

struct Celula *novoCelula(Personagem elemento) {
    struct Celula *novoCelula = (struct Celula *)malloc(sizeof(struct Celula));
    if (novoCelula == NULL) {
        exit(EXIT_FAILURE);
    }
    novoCelula->elemento = elemento;
    novoCelula->prox = NULL;
    return novoCelula;
}

void start(tabelaHash *tabela) {
    tabela->cabeca = NULL;
    tabela->count = 0;
}

void inserir(tabelaHash *tabela, Personagem elemento) {
    Celula *novo = novoCelula(elemento);
    novo->prox = tabela->cabeca;
    tabela->cabeca = novo;
    tabela->count=tabela->count++;
}

bool pesquisar(tabelaHash *tabela, char* elemento,int pos) {    

    Celula *i = tabela->cabeca;
    while (i != NULL) {
        comp++;
        if (strcmp(i->elemento.name, elemento) == 0) {           
            printf(" (pos: %d)",pos);
            return true;
        }
        i = i->prox;
    }
    return false;
}

// Questao 7
void Questao7(Personagem personagens[])
{
    tabelaHash tabelahash[21];
    char aux[200];


    for (int i = 0; i < 21; i++) {
        start(&tabelahash[i]);
    }

    fgets(aux, 200, stdin); 
    aux[strcspn(aux, "\n\r")] = '\0';  
    
    while (strcmp(aux, "FIM") != 0)
    {
        for (int i = 0; i < 405; i++)
        {
            if (strcmp(personagens[i].id, aux) == 0)
            {
                int valor = 0;
                for (int j = 0; j < strlen(personagens[i].name); j++) {
                    valor += personagens[i].name[j];
                }

                int pos = valor % 21;
                inserir(&tabelahash[pos], personagens[i]);
                i = 500;
            }
        }
        fgets(aux, 200, stdin); 
        aux[strcspn(aux, "\n\r")] = '\0'; 
        
    }
    fgets(aux,200,stdin);    
    aux[strcspn(aux, "\n\r")] = '\0'; 
    
    
    while (strcmp(aux, "FIM") != 0)
    {  
        printf("%s", aux);
        int valor = 0;
        for (int i = 0; i < strlen(aux); i++) {
            valor += aux[i];
        }
        int pos = valor % 21;   

        bool resp = pesquisar(&tabelahash[pos],aux,pos);  
            
        printf(" %s\n", resp ? "SIM" : "NAO");
        fgets(aux, 200, stdin); 
        aux[strcspn(aux, "\n\r")] = '\0'; 
        
    }
}

void criaArqLog() {
    FILE* arqLog;
    arqLog = fopen("matricula_hashIndireta.txt", "w");

    if (arqLog == NULL) {
        printf("Erro ao criar o arquivo de log\n");
        return;
    }
    int totalTime = ((double)(endTime - startTime)) / CLOCKS_PER_SEC;
    fprintf(arqLog, "126920\t%.2f\t%d\n",totalTime, comp);
    fclose(arqLog);
} 

int main()
{
    startTimer();
    Personagem personagens[405];
    preencherVetor(personagens);
    Questao7(personagens);
    endTimer();
    criaArqLog();
    return 0;
}